File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/BalanceReadingInputDialog.java
package com.gb.canibuythat.ui;

import android.app.AlertDialog;
import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.gb.canibuythat.R;
import com.gb.canibuythat.UserPreferences;
import com.gb.canibuythat.di.Injector;
import com.gb.canibuythat.ui.model.BalanceReading;
import com.gb.canibuythat.util.DateUtils;
import com.gb.canibuythat.util.ViewUtils;

import java.util.Date;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

public class BalanceReadingInputDialog extends DialogFragment
        implements View.OnClickListener {

    @Inject UserPreferences userPreferences;

    private BalanceReading lastUpdate;

    @BindView(R.id.last_update) TextView lastUpdateView;
    @BindView(R.id.amount) EditText valueView;
    @BindView(R.id.when_btn) DatePickerButton whenButton;

    private Unbinder unbinder;

    public BalanceReadingInputDialog() {
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Injector.INSTANCE.getGraph().inject(this);
        lastUpdate = userPreferences.getBalanceReading();
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        LinearLayout body = (LinearLayout) LayoutInflater.from(getActivity()).inflate(R.layout.dialog_balance_reading, null);
        unbinder = ButterKnife.bind(this, body);

        Date today = new Date();
        whenButton.setText(DateUtils.getFORMAT_MONTH_DAY_YR().format(today));
        whenButton.setDate(today);

        if (lastUpdate != null) {
            valueView.setText(Float.toString(lastUpdate.balance));
            lastUpdateView.setText(getString(R.string.balance_update_reading, lastUpdate.balance,
                    DateUtils.getFORMAT_MONTH_DAY_YR().format(lastUpdate.when)));
        } else {
            lastUpdateView.setText("None");
        }

        return new AlertDialog.Builder(getActivity()).setTitle("Set starting balance")
                .setPositiveButton(getText(android.R.string.ok), null)
                .setNegativeButton(getText(android.R.string.cancel), null)
                .setView(body)
                .create();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onStart() {
        super.onStart();
        final AlertDialog dialog = (AlertDialog) getDialog();

        if (dialog != null) {
            Button b = dialog.getButton(AlertDialog.BUTTON_POSITIVE);
            b.setOnClickListener(this);
        }
        ViewUtils.showKeyboard(valueView);
    }

    @Override
    public void onClick(View view) {
        if (validate()) {
            Date selectedDate = whenButton.getSelectedDate();
            BalanceReading balanceReading = new BalanceReading(DateUtils.clearLowerBits(selectedDate),
                    Float.valueOf(valueView.getText().toString()));
            userPreferences.setBalanceReading(balanceReading);
            dismiss();
        }
    }

    private boolean validate() {
        if (TextUtils.isEmpty(valueView.getText())) {
            valueView.setError("Please specify an amount!");
            valueView.requestFocus();
            return false;
        }

        Date selectedDate = whenButton.getSelectedDate();
        if (!selectedDate.before(new Date())) {
            Toast.makeText(getActivity(), "Please select a non-future date!", Toast.LENGTH_SHORT).show();
            return false;
        }

        Date estimateDate = userPreferences.getEstimateDate();
        if (selectedDate.after(estimateDate)) {
            Toast.makeText(getActivity(), "Please select a date before the target date!", Toast.LENGTH_SHORT).show();
            return false;
        }
        return true;
    }
}

File: app_canibuythat/src/main/java/com/gb/canibuythat/ui/BalanceReadingInputDialog.kt
package com.gb.canibuythat.ui

import android.app.AlertDialog
import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.text.TextUtils
import android.view.LayoutInflater
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.Toast
import com.gb.canibuythat.R
import com.gb.canibuythat.UserPreferences
import com.gb.canibuythat.di.Injector
import com.gb.canibuythat.ui.model.BalanceReading
import com.gb.canibuythat.util.DateUtils
import com.gb.canibuythat.util.clearLowerBits
import com.gb.canibuythat.util.showKeyboard
import java.util.*
import javax.inject.Inject

class BalanceReadingInputDialog : DialogFragment(), DialogInterface.OnClickListener {

    @Inject lateinit var userPreferences: UserPreferences
    private var lastUpdate: BalanceReading? = null

    lateinit var body: LinearLayout
    val amountInput: EditText by lazy { body.findViewById(R.id.amount_input) as EditText }
    val whenBtn: DatePickerButton by lazy { body.findViewById(R.id.when_btn) as DatePickerButton }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Injector.INSTANCE.graph.inject(this)
        lastUpdate = userPreferences.balanceReading
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        body = LayoutInflater.from(activity).inflate(R.layout.dialog_balance_reading, null) as LinearLayout

        lastUpdate?.let {
            amountInput.setText(java.lang.Float.toString(lastUpdate!!.balance))
            whenBtn.text = DateUtils.formatDayMonthYear(it.`when`!!)
            whenBtn.setDate(it.`when`)
        } ?: let {
            val today = Date()
            whenBtn.text = DateUtils.formatDayMonthYearWithPrefix(today)
            whenBtn.setDate(today)
        }

        return AlertDialog.Builder(activity).setTitle("Set starting balance")
                .setPositiveButton(getText(android.R.string.ok), this)
                .setNegativeButton(getText(android.R.string.cancel), null)
                .setView(body)
                .create()
    }

    override fun onStart() {
        super.onStart()
        amountInput.showKeyboard()
    }

    override fun onClick(dialog: DialogInterface?, which: Int) {
        if (validate()) {
            val selectedDate = whenBtn.selectedDate
            userPreferences.balanceReading = BalanceReading(selectedDate.clearLowerBits(),
                    amountInput.text.toString().toFloat())
            dismiss()
        }
    }

    private fun validate(): Boolean {
        if (TextUtils.isEmpty(amountInput.text)) {
            amountInput.error = "Please specify an amount!"
            amountInput.requestFocus()
            return false
        }

        val selectedDate = whenBtn.selectedDate
        if (!selectedDate.before(Date())) {
            Toast.makeText(activity, "Please select a non-future date!", Toast.LENGTH_SHORT).show()
            return false
        }

        val estimateDate = userPreferences.estimateDate
        if (selectedDate.after(estimateDate)) {
            Toast.makeText(activity, "Please select a date before the target date!", Toast.LENGTH_SHORT).show()
            return false
        }
        return true
    }
}
